File: Management/Models/Local/LocalModules.cs
namespace LocalModels;

public sealed record LocalModule
{
  public string Name { get; init; } = string.Empty;
  public string Notes { get; set; } = string.Empty;
  public IEnumerable<LocalAssignment> Assignments { get; init; } = [];
  public IEnumerable<LocalQuiz> Quizzes { get; init; } = [];
  public IEnumerable<LocalCoursePage> Pages { get; init; } = [];

  public IEnumerable<IModuleItem> GetSortedModuleItems() =>
    Enumerable.Empty<IModuleItem>()
      .Concat(Assignments)
      .Concat(Quizzes)
      .Concat(Pages)
      .OrderBy(i => i.DueAt);

  public bool Equals(LocalModule? otherModule)
  {
    var areEqual =
      string.Equals(Name, otherModule?.Name, StringComparison.OrdinalIgnoreCase)
      && string.Equals(Notes, otherModule?.Notes, StringComparison.OrdinalIgnoreCase)
      && CompareCollections(Assignments.OrderBy(x => x.Name), otherModule?.Assignments.OrderBy(x => x.Name))
      && CompareCollections(Quizzes.OrderBy(x => x.Name), otherModule?.Quizzes.OrderBy(x => x.Name))
      && CompareCollections(Pages.OrderBy(x => x.Name), otherModule?.Pages.OrderBy(x => x.Name));
    return areEqual;
  }

  private static bool CompareCollections<T>(IEnumerable<T> first, IEnumerable<T>? second)
  {
    var firstList = first.ToList();
    var secondList = second?.ToList();

    if (firstList.Count != secondList?.Count)
      return false;

    for (int i = 0; i < firstList.Count; i++)
    {
      if (!Equals(firstList[i], secondList[i]))
        return false;
    }

    return true;
  }

  public override int GetHashCode()
  {
    HashCode hash = new HashCode();
    hash.Add(Name, StringComparer.OrdinalIgnoreCase);
    hash.Add(Notes, StringComparer.OrdinalIgnoreCase);
    AddRangeToHash(hash, Assignments.OrderBy(x => x.Name));
    AddRangeToHash(hash, Quizzes.OrderBy(x => x.Name));
    AddRangeToHash(hash, Pages.OrderBy(x => x.Name));
    return hash.ToHashCode();
  }

  private void AddRangeToHash<T>(HashCode hash, IEnumerable<T> items)
  {
    foreach (var item in items)
    {
      hash.Add(item);
    }
  }
}

File: nextjs/src/models/local/localModules.ts
import { LocalAssignment } from "./assignmnet/localAssignment";
import { IModuleItem } from "./IModuleItem";
import { LocalCoursePage } from "./page/localCoursePage";
import { LocalQuiz } from "./quiz/localQuiz";
import { getDateFromString } from "./timeUtils";

export interface LocalModule {
  name: string;
  assignments: LocalAssignment[];
  quizzes: LocalQuiz[];
  pages: LocalCoursePage[];
}

export const LocalModuleUtils = {
  getSortedModuleItems(module: LocalModule): IModuleItem[] {
    return [...module.assignments, ...module.quizzes, ...module.pages].sort(
      (a, b) =>
        (getDateFromString(a.dueAt)?.getTime() ?? 0) -
        (getDateFromString(b.dueAt)?.getTime() ?? 0)
    );
  },

  equals(module1: LocalModule, module2: LocalModule): boolean {
    return (
      module1.name.toLowerCase() === module2.name.toLowerCase() &&
      LocalModuleUtils.compareCollections(
        module1.assignments.sort((a, b) => a.name.localeCompare(b.name)),
        module2.assignments.sort((a, b) => a.name.localeCompare(b.name))
      ) &&
      LocalModuleUtils.compareCollections(
        module1.quizzes.sort((a, b) => a.name.localeCompare(b.name)),
        module2.quizzes.sort((a, b) => a.name.localeCompare(b.name))
      ) &&
      LocalModuleUtils.compareCollections(
        module1.pages.sort((a, b) => a.name.localeCompare(b.name)),
        module2.pages.sort((a, b) => a.name.localeCompare(b.name))
      )
    );
  },

  compareCollections<T>(first: T[], second: T[]): boolean {
    if (first.length !== second.length) return false;

    for (let i = 0; i < first.length; i++) {
      if (JSON.stringify(first[i]) !== JSON.stringify(second[i])) return false;
    }

    return true;
  },

  getHashCode(module: LocalModule): number {
    const hash = new Map<string, number>();
    hash.set(module.name.toLowerCase(), 1);
    LocalModuleUtils.addRangeToHash(
      hash,
      module.assignments.sort((a, b) => a.name.localeCompare(b.name))
    );
    LocalModuleUtils.addRangeToHash(
      hash,
      module.quizzes.sort((a, b) => a.name.localeCompare(b.name))
    );
    LocalModuleUtils.addRangeToHash(
      hash,
      module.pages.sort((a, b) => a.name.localeCompare(b.name))
    );

    return Array.from(hash.values()).reduce((acc, val) => acc + val, 0);
  },

  addRangeToHash<T>(hash: Map<string, number>, items: T[]): void {
    for (const item of items) {
      hash.set(JSON.stringify(item), 1);
    }
  },
};
